onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selectors.cc

```diff
@@ -330,6 +330,30 @@ bool WhereNodeGroupSelector::Check(const GraphViewer& graph_viewer, const Node&
          dt_input_1 == dt_output;
 }
 
+bool PadNodeGroupSelector::Check(const GraphViewer& graph_viewer, const Node& node,
+                                 const std::vector<const Node*>& dq_nodes,
+                                 const std::vector<const Node*>& q_nodes) const {
+  int num_outputs = NumActualValues(node, false);  // number of outputs that exist
+  bool output_check = (num_outputs == gsl::narrow_cast<int>(q_nodes.size())) &&
+                      q_nodes.size() == node.GetOutputEdgesCount() &&
+                      !graph_viewer.NodeProducesGraphOutput(node);
+  if (!output_check) {
+    return false;
+  }
```

Comment:
It seems to me that the above 7 lines could be replaced with a call to `CheckQDQNodes(graph_viewer, node, dq_nodes, q_nodes, -1)`.  In fact, the use of `CheckQDQNodes` would also validate the DQ nodes, which this is missing.

But maybe I'm missing something.

Reply:
That's something I'm trying to avoid here. Pad has 3 inputs: data, pads, constant_value. pads is not quantized. constant_value can be quantized or non-quantized, Qnn both support. The code below checks dq node for input[0] with q node for output to make sure they have same type. and if dq exist for constant_value, it should has same type with input[0].

---
In reply to: [1325165768](https://github.com/microsoft/onnxruntime/pull/17508#discussion_r1325165768) [](ancestors = 1325165768)

Reply:
I see. That explains why we can't use `CheckQDQNodes` directly. 

However, we should still call `QDQ::ValidateNodeGroupDQNodes` on each quantized input to ensure that the DQ does not produce a graph output and has a single output. If the second input is quantized, then we should call `QDQ::ValidateNodeGroupDQNodes` on it too.

Reply:
Yes, will update it.


---
In reply to: [1325412056](https://github.com/microsoft/onnxruntime/pull/17508#discussion_r1325412056) [](ancestors = 1325412056)